### ISSUE TITLE
Broadcasting Assigments In a Loop

### ISSUE THREAD

**Issue by glenflet** [numpy/numpy#8802]
This isn't exactly a bug in numpy, rather a source of hard to trace bugs in code written in numpy

In code such as:
```.py
data = np.empty((numlines,) dtype=structure)
for i in range(numlines):
    line = # Read line from input source
    ...
    data[field][i,:] = line[start:end]
    ...
```
because of numpys broadcasting rules
```.py
data = np.empty((numlines,) dtype=structure)
for i in range(numlines):
    line = # Read line from input source
    ...
    data[field][:] = line[start:end]
    ...
```
is valid code, and a easy typo, the result is that the field is overwritten on each iteration of the loop rather than simply adding the value at the index, and all values appear to be the same.

Obviously the broadcasting rules can't be change but, is it possible to cause such code to print a warning to stderr the first time its evaluated as this would make tracing such errors easier, this would of course be best if we could make it context aware, however I'm not sure if that is possible.

This exact typo has, recently caused me a loot of grief in tracking down a bug in some code I was working with, as the code looked right, when quickly skimming over it I end up first looking at the files I was reading believe the error to be there and only found this error after adding some debug print statement. On the other hand if my loading function produced a message such as "WARNING: Broadcasting assignment to array" I would have taken a closer look at the assignment and found this bug in a matter of minutes instead of wasting a couple of hours trying to debug the wrong part of the code base.

**Comment by njsmith** [comment]
If we were going to do something about this, I think the way to do it would be to issue a warning when using an "incomplete" index, so e.g. when indexing a 2d array you would explicitly have to say `arr[i, ...]` instead of `arr[i]`. I'm doubtful that we could actually get away with this though... while it's error prone and (IMO) encourages bad habits, my experience is that a *lot* of our userbase uses the `arr[i]` style and this would be pretty disruptive. There's probably some other issue discussing this already...

**Comment by glenflet** [comment]
You would only need a message when broadcasting to an incomplete index, and mostly only when the provide dimension match the number of dimensions used to index.

say let aN be an Nd array then

```.py
a3[i,:] = a1 # Warning Specified 1 range index, but has an implicit extra which cause broadcasting
a3[i] = a1 # No Error, since no range index given
a3[i,...] = a1 # No Error, any number of dimensions excepted, broadcasting okay
a3[i,:,:] = a1 # No Error, Explicit Specification of dimensions, broadcasting okay
a3[:] = a1 # Warning Specified 1 range index, but has an implicit extra which cause broadcasting
a3[...] = a1 # No Error, any number of dimensions excepted, broadcasting okay
a3[:,:] = a1 # Warning Specified 2 range index, but has an implicit extra which cause broadcasting
a3[:,:,:] = a1 # No Error, Explicit Specification of dimensions, broadcasting okay
```

This would be unlike to cause many problem as if a user is being explicit with the dimensions as shown they can be reasonable expected to be explicit with all dimensions 

and the trackback module could be used to locate the call site, hence the user would get a warning like

```WARNING: Broadcasting to Incorrect number of explicit Dimensions, on line X in file Y```

a internal flag could be used so the warning would only be printed once per call site, and a global flag could be used to suppress the warning

It may even be possible to only log the first invocation of a call site, and print the error on the second there by detecting the usage only in loops which would imply a missing index variable

**Comment by eric-wieser** [comment]
I do think there's some value in here somewhere, but I don't think this is a good idea as you've suggested it so far. An example of when I don't want this:

    img = np.zeros((32, 32, 3), dtype=np.uint8)
    red = np.array([255, 0, 0]. dtype=np.uint8)
    img[:,:] = red  # set every pixel of the "2d" image to red

> say let aN be an Nd array then

What about `a4[:,i], a4[:,:,i]`? That should warn if you treat it the same as `a4[i,:], a4[i,:,:]` but should not if you treat it as an extension of `a4[i]`.

> and the trackback module could be used to locate the call site [...] a internal flag could be used so the warning would only be printed once per call site

I think the builtin `warnings` module already knows how to handle this.

**Comment by glenflet** [comment]
>> say let aN be an Nd array then
>
> What about `a4[:,i], a4[:,:,i]`? That should warn if you treat it the same as `a4[i,:], a4[i,:,:]` but should not if you treat it as an extension of `a4[i]`.

Would be the same as `a4[i,:], a4[i,:,:]` since the my reasoning there was that if the user provide any non numerical index, we assume they mean to be explicit in their indexing, the position on any provided numerical index's wouldn't matter.

As for you example of where you wouldn't want this I can see you point however, its also exactly equivalent to my initial problem just a different number of overall shape. On the other hand would a single warning written to stderr really be much of a problem here, since it wouldn't actually be breaking the code, just letting the user know there may be something wrong.

Possible solutions would be a flag variable to decide how to teat such situations and as I previously suggested possibly with detecting call site, only produce a warning when repeated invocation overwrites previous data exactly, as this is indicative of an missing index.

**Comment by pv** [comment]
I don't see incomplete indices as bad style TBH.
.
I think the numpy indexing semantics are for the better or worse mostly 
fixed at this point, and changes to make them slightly nicer in some 
subjective way have a net negative impact.

**Comment by seberg** [comment]
I agree. When thinking about `oindex`, etc. I thought that those should not allow incomplete indices (because in principle I half agree that most incomplete indices are easy to avoid in new code).

I could (a bit) imaging a deprecation for tuple/non-singleton indices, only. That would mean `arr3d[1]` or `arr3d[1:4]` would be fine, but `arr3d[1,]` or `arr3d[1:4, :]` would warn/error about incomplete index.

**Comment by eric-wieser** [comment]
> On the other hand would a single warning written to stderr really be much of a problem here

Yes. Warnings can cause users to change their code to make the warning go away. We should never encourage users to change their code to something less clear, which I believe specifying the last dimension on an image is.

I think fundamentally, the issue is that sometime dimensions form part of the "element", and not the array. For instance, in an image, the base element is really of shape (3,), and most code wants to think about it in terms of a 2d array. In these cases, the warning is not desirable

But there're more cases than that - we also need to deal with things like stacks of images, where a shape of `(N, W, H, 3)` is really better interpreted as `(N,) x (W,H) x (3,)`. Ideally, we'd issue a warnings when one of these dimension groups is not fully indexed - but right now, the only way to represent this is the incredibly clumsy, but thankfully not that costly:

```
dt = np.dtype([('image', np.dtype([('pixel', uint8, (3,))]), (W, H)])
```

**Comment by pv** [comment]
To put it more officially, -1 from me to warnings/etc for incomplete 
indices.
.
I think changes that break code should have good justifications --- I 
think "it's ugly" and "it's a bit misleading" are problematic reasons to 
make changes in basic language semantics.
.
As a N=1 case study, I submit that last time I fixed code written two 
years ago broken in the meantime by a Numpy change, was today. My 
colleague who I was sending it to would not have been able to fix it.

**Comment by eric-wieser** [comment]
> I think changes that break code should have good justifications 

I'd argue that warnings aren't breaking code, but I think i'm at best +0 on this anyway.

> broken in the meantime by a Numpy change

What was the change?

**Comment by seberg** [comment]
Don't underestimate how intimidating a warning that is not understood can be a for a new user. We have been doing a lot of things that have caused pain downstream, this one would be extremely visible as well. And frankly, I doubt we can agree on it anyway ;). We also have to admit that we have been getting pretty "fast" with larger changes in the past years. Sure, I like to think we always had very good reasons, but the fact is also that we are not the ones who get hit by the problems really, so its hard to estimate the pain/gain trade-off.
